sky/packages/sky/lib/gestures/recognizer.dart

```diff
@@ -0,0 +1,141 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:sky' as sky;
+
+import 'package:sky/base/pointer_router.dart';
+import 'package:sky/gestures/arena.dart';
+import 'package:sky/gestures/constants.dart';
+
+abstract class GestureRecognizer extends GestureArenaMember {
+  GestureRecognizer({ PointerRouter router }) : _router = router;
+
+  PointerRouter _router;
+
+  final List<GestureArenaEntry> _entries = new List<GestureArenaEntry>();
+  final Set<int> _trackedPointers = new Set<int>();
+
+  void addPointer(sky.PointerEvent event);
+  void handleEvent(sky.PointerEvent event);
+  void didStopTrackingLastPointer();
+
+  void resolve(GestureDisposition disposition) {
+    List<GestureArenaEntry> localEntries = new List.from(_entries);
+    _entries.clear();
+    for (GestureArenaEntry entry in localEntries)
+      entry.resolve(disposition);
+  }
+
+  void dispose() {
+    resolve(GestureDisposition.rejected);
+    for (int pointer in _trackedPointers)
+      _router.removeRoute(pointer, handleEvent);
+    _trackedPointers.clear();
+    assert(_entries.isEmpty);
+    _router = null;
+  }
+
+  void startTrackingPointer(int pointer) {
+    _router.addRoute(pointer, handleEvent);
+    _trackedPointers.add(pointer);
+    _entries.add(GestureArena.instance.add(pointer, this));
+  }
+
+  void stopTrackingPointer(int pointer) {
+    _router.removeRoute(pointer, handleEvent);
+    _trackedPointers.remove(pointer);
+    if (_trackedPointers.isEmpty)
+      didStopTrackingLastPointer();
+  }
+
+  void stopTrackingIfNotDown(sky.PointerEvent event) {
```

Comment:
It's not obvious if this is stopping if the _event_ isn't down, or if the _pointer_ isn't down.